core/io/multiplayer_peer_custom.cpp

```diff
@@ -0,0 +1,184 @@
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "core/io/multiplayer_peer_custom.h"
+
+MultiplayerPeerCustom::MultiplayerPeerCustom() {
+	self_id = 0;
+	transfer_mode = TransferMode::TRANSFER_MODE_RELIABLE;
+	connection_status = ConnectionStatus::CONNECTION_DISCONNECTED;
+	refusing_new_connections = false;
+	target_id = 0;
+	// Default to a large value.
+	max_packet_size = 1 << 24;
+}
+
+MultiplayerPeerCustom::~MultiplayerPeerCustom() {
+}
+
+void MultiplayerPeerCustom::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("initialize", "self_peer_id"), &MultiplayerPeerCustom::initialize);
+	ClassDB::bind_method(D_METHOD("set_max_packet_size", "max_packet_size"), &MultiplayerPeerCustom::set_max_packet_size);
+	ClassDB::bind_method(D_METHOD("set_connection_status", "connection_status"), &MultiplayerPeerCustom::set_connection_status);
+	ClassDB::bind_method(D_METHOD("deliver_rpc", "buffer", "from_peer_id"), &MultiplayerPeerCustom::deliver_rpc);
+
+	ADD_SIGNAL(MethodInfo("rpc_generated", PropertyInfo(Variant::INT, "peer_id"), PropertyInfo(Variant::POOL_BYTE_ARRAY, "buffer"), PropertyInfo(Variant::INT, "transfer_mode")));
+}
+
+//
+// PacketPeer
+//
+
+Error MultiplayerPeerCustom::get_packet(const uint8_t **r_buffer, int &r_buffer_size) {
+	ERR_FAIL_COND_V(incoming_packets.size() == 0, Error::ERR_UNAVAILABLE);
+
+	current_packet = incoming_packets.front()->get();
+	incoming_packets.pop_front();
+
+	*r_buffer = current_packet.data.read().ptr();
+	r_buffer_size = current_packet.data.size();
+
+	return Error::OK;
+}
+
+Error MultiplayerPeerCustom::put_packet(const uint8_t *p_buffer, int p_buffer_size) {
+	PoolByteArray a;
+	a.resize(p_buffer_size);
+
+	PoolByteArray::Write w = a.write();
+	memcpy(w.ptr(), p_buffer, p_buffer_size);
+
+	emit_signal("rpc_generated", target_id, a, transfer_mode);
```

Comment:
Since the packet content might not be an RPC (e.g. it might be raw bytes, a path cache simplify/ack, etc: https://github.com/godotengine/godot/blob/3.x/core/io/multiplayer_api.cpp#L189-L198 ) maybe this should be more consistently named `packet_generated`